File: app/api/src/main/kotlin/com/wallet/clover/api/application/TicketService.kt
package com.wallet.clover.api.application

import com.wallet.clover.adapter.LottoTicketClient
import com.wallet.clover.domain.game.LottoGame
import com.wallet.clover.domain.ticket.LottoTicket
import com.wallet.clover.domain.ticket.SaveScannedTicketCommand
import com.wallet.clover.domain.ticket.parser.TicketParser
import com.wallet.clover.repository.game.LottoGameRdbAdaptor
import com.wallet.clover.repository.ticket.LottoTicketRdbAdaptor
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
@Transactional(readOnly = true)
class TicketService(
    val lottoTicketRdbAdaptor: LottoTicketRdbAdaptor,
    val lottoGameRdbAdaptor: LottoGameRdbAdaptor,
    val lottoTicketFeignClient: LottoTicketClient,
    val ticketParser: TicketParser,
) {
    @Transactional
    fun saveScannedTicket(command: SaveScannedTicketCommand): LottoTicket {
        val preRegistered = lottoTicketRdbAdaptor.byUserIdAndUrl(command.userId, command.url)
        if (preRegistered != null) {
            return preRegistered
        }

        val html = lottoTicketFeignClient.getHtmlByUrl(command.url)
        val parsedTicket = ticketParser.parse(html)

        val ticket = lottoTicketRdbAdaptor.saveImmediately(
            LottoTicket(
                userId = command.userId,
                url = command.url,
                ordinal = parsedTicket.ordinal,
                status = parsedTicket.status,
            ),
        )

        val games = parsedTicket.games.map {
            LottoGame(
                ticketId = ticket.id,
                userId = command.userId,
                status = it.status,
                number1 = it.number1,
                number2 = it.number2,
                number3 = it.number3,
                number4 = it.number4,
                number5 = it.number5,
                number6 = it.number6,
            )
        }
        lottoGameRdbAdaptor.saveAll(games)
        return ticket
    }

    fun byUserId(userId: Long): List<LottoTicket> {
        return lottoTicketRdbAdaptor.byUserId(userId)
    }

    fun byId(ticketId: Long): LottoTicket {
        return lottoTicketRdbAdaptor.byId(ticketId)
    }

    fun getGamesByTicketId(ticketId: Long): List<LottoGame> {
        return lottoGameRdbAdaptor.byTicketId(ticketId)
    }
}

File: apps/backend/src/ticket/ticket.service.ts
import { Injectable, NotFoundException, Logger } from '@nestjs/common';
import { PrismaService } from '../prisma/prisma.service';
import { LottoTicketClient } from './client/lotto-ticket.client';
import { JsoupTicketParser } from './client/jsoup-ticket.parser';

@Injectable()
export class TicketService {
  private readonly logger = new Logger(TicketService.name);

  constructor(
    private readonly prisma: PrismaService,
    private readonly lottoTicketClient: LottoTicketClient,
    private readonly ticketParser: JsoupTicketParser,
  ) {}

  async getMyTickets(userId: bigint, page = 0, size = 20) {
    const skip = page * size;
    const [content, total] = await Promise.all([
      this.prisma.lottoTicket.findMany({
        where: { userId },
        skip,
        take: size,
        orderBy: { createdAt: 'desc' },
      }),
      this.prisma.lottoTicket.count({ where: { userId } }),
    ]);

    return {
      content,
      page,
      size,
      totalElements: total,
      totalPages: Math.ceil(total / size),
    };
  }

  async getTicketById(ticketId: bigint) {
    const ticket = await this.prisma.lottoTicket.findUnique({
      where: { id: ticketId },
      include: { games: true },
    });

    if (!ticket) {
      throw new NotFoundException(`티켓을 찾을 수 없습니다: ${ticketId}`);
    }

    return ticket;
  }

  async saveScannedTicket(
    userId: bigint,
    url: string,
    extractionMethod?: string,
  ) {
    // 1. 이미 등록된 티켓인지 확인
    const existingTicket = await this.prisma.lottoTicket.findUnique({
      where: { url },
    });

    if (existingTicket) {
      this.logger.log(`이미 등록된 티켓입니다: ${url}`);
      return existingTicket;
    }

    try {
      // 2. 외부 사이트에서 HTML 가져오기 및 파싱
      const html = await this.lottoTicketClient.getHtmlByUrl(url);
      const parsedTicket = this.ticketParser.parse(html);

      // 3. 트랜잭션으로 티켓과 게임 정보 저장
      return await this.prisma.$transaction(async (tx) => {
        const savedTicket = await tx.lottoTicket.create({
          data: {
            userId,
            url,
            ordinal: parsedTicket.ordinal,
            status: parsedTicket.status,
            createdAt: new Date(),
            updatedAt: new Date(),
          },
        });

        const gameData = parsedTicket.games.map((g) => ({
          ticketId: savedTicket.id,
          userId,
          status: g.status,
          number1: g.number1,
          number2: g.number2,
          number3: g.number3,
          number4: g.number4,
          number5: g.number5,
          number6: g.number6,
          extractionMethod: extractionMethod || 'QR_SCAN',
          createdAt: new Date(),
          updatedAt: new Date(),
        }));

        await tx.lottoGame.createMany({
          data: gameData,
        });

        return savedTicket;
      });
    } catch (error) {
      this.logger.error(
        `티켓 스캔 및 저장 중 오류 발생: ${error.message}`,
        error.stack,
      );
      throw error;
    }
  }
}
